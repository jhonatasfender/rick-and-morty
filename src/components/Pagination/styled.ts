import styled, { css } from 'styled-components'

import { NavigateBefore, NavigateNext } from '@styled-icons/material-rounded'

export const Wrapper = styled.ul`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    height: 10rem;
    gap: 1rem;
    width: 100%;
    list-style-type: none;
  `}
`

type ItemProps = {
  active:boolean
}

export const Item = styled.li<ItemProps>`
  ${({ theme,active }) => css`
    --size: ${active? '5rem':'4rem'};
  
    background-color: ${theme.colors.button};
    font-size: 1.8rem;
    font-weight: bold;
    border-radius: 50%;
    width: var(--size);
    height: var(--size);
    text-align: center;
    line-height: var(--size);
    cursor: pointer;
    user-select: none;
  `}
`

export const Prev = styled(NavigateBefore)`
  ${({ theme }) => css`
    width: 8rem;
    fill: ${theme.colors.button};
    cursor: pointer;
  `}
`

export const Next = styled(NavigateNext)`
  ${({ theme }) => css`
    width: 8rem;
    fill: ${theme.colors.button};
    cursor: pointer;
  `}
`
