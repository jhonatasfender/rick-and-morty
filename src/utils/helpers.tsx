import { ReactNode } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

import { render, RenderResult } from '@testing-library/react'

import theme from '../styles/theme'

const queryClient = new QueryClient()

export const renderWithThemeAndRoute = (children: ReactNode): RenderResult =>
  render(
    <QueryClientProvider client={queryClient}>
      <MemoryRouter>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </MemoryRouter>
    </QueryClientProvider>,
  )
