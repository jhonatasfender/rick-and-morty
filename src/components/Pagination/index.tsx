import { Dispatch, SetStateAction } from 'react'

import * as S from './styled'

type PaginationProps = {
  renge: string[]
  page: number
  total?: number
  setPage: Dispatch<SetStateAction<number>>
}

export default function Pagination({
  renge,
  page,
  setPage,
  total,
}: PaginationProps): JSX.Element {
  const handlePrev = () => {
    setPage((prev) => (prev <= 1 ? prev : prev - 1))
  }

  const handleNext = () => {
    setPage((prev) => (total === prev ? prev : prev + 1))
  }

  const handleSpecific = (current: string) => {
    if (current !== '...') {
      setPage(Number(current))
    }
  }

  return (
    <S.Wrapper aria-label="Pagination Navigation">
      <S.Prev onClick={handlePrev} aria-hidden="false" />
      {renge.map((num, key) => (
        <S.Item
          active={page.toString() === num}
          key={key}
          onClick={() => handleSpecific(num)}
        >
          {num}
        </S.Item>
      ))}
      <S.Next onClick={handleNext} aria-hidden="false" />
    </S.Wrapper>
  )
}
