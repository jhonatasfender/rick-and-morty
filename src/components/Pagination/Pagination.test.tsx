import { renderWithThemeAndRoute } from '../../utils/helpers'
import Pagination from '.'

describe('<Pagination />', () => {
  it('should load the component with default values', () => {
    const { container } = renderWithThemeAndRoute(
      <Pagination
        page={1}
        setPage={jest.fn()}
        renge={['1', '2', '3', '4', '5']}
      />,
    )

    expect(container).toMatchSnapshot()
  })
})
