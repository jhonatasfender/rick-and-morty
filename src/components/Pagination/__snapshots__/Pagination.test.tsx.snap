// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should load the component with default values 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  color: #ffffff;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 10rem;
  gap: 1rem;
  width: 100%;
  list-style-type: none;
}

.c3 {
  --size: 5rem;
  background-color: rgb(157 23 30);
  font-size: 1.8rem;
  font-weight: bold;
  border-radius: 50%;
  width: var(--size);
  height: var(--size);
  text-align: center;
  line-height: var(--size);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  --size: 4rem;
  background-color: rgb(157 23 30);
  font-size: 1.8rem;
  font-weight: bold;
  border-radius: 50%;
  width: var(--size);
  height: var(--size);
  text-align: center;
  line-height: var(--size);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  width: 8rem;
  fill: rgb(157 23 30);
  cursor: pointer;
}

.c5 {
  width: 8rem;
  fill: rgb(157 23 30);
  cursor: pointer;
}

<div>
  <ul
    aria-label="Pagination Navigation"
    class="c0"
  >
    <svg
      aria-hidden="false"
      class="c1 c2"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M14.91 6.71a.996.996 0 0 0-1.41 0L8.91 11.3a.996.996 0 0 0 0 1.41l4.59 4.59a.996.996 0 1 0 1.41-1.41L11.03 12l3.88-3.88c.38-.39.38-1.03 0-1.41z"
      />
    </svg>
    <li
      class="c3"
    >
      1
    </li>
    <li
      class="c4"
    >
      2
    </li>
    <li
      class="c4"
    >
      3
    </li>
    <li
      class="c4"
    >
      4
    </li>
    <li
      class="c4"
    >
      5
    </li>
    <svg
      aria-hidden="false"
      class="c1 c5"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9.31 6.71a.996.996 0 0 0 0 1.41L13.19 12l-3.88 3.88a.996.996 0 1 0 1.41 1.41l4.59-4.59a.996.996 0 0 0 0-1.41L10.72 6.7c-.38-.38-1.02-.38-1.41.01z"
      />
    </svg>
  </ul>
</div>
`;
